homeassistant/components/roborock/image.py

```diff
@@ -163,32 +181,55 @@ async def create_coordinator_maps(
-            coord.current_map = map_flag
-            # We cannot get the map until the roborock servers fully process the
-            # map change.
-            await asyncio.sleep(MAP_SLEEP)
-        # Get the map data
-        map_update = await asyncio.gather(
-            *[coord.cloud_api.get_map_v1(), coord.get_rooms()], return_exceptions=True
+    maps = await hass.async_add_executor_job(
+        roborock_storage.exec_load_maps,
+        [roborock_map.name for roborock_map in coord.maps.values()],
+    )
+    storage_updates: list[tuple[str, bytes]] = []
+    for (map_flag, map_info), storage_map in zip(maps_info, maps, strict=False):
+        unique_id = (
+            f"{slugify(coord.roborock_device_info.device.duid)}_map_{map_info.name}"
         )
-        # If we fail to get the map, we should set it to empty byte,
-        # still create it, and set it as unavailable.
-        api_data: bytes = map_update[0] if isinstance(map_update[0], bytes) else b""
-        entities.append(
-            RoborockMap(
-                f"{slugify(coord.duid)}_map_{map_info.name}",
-                coord,
-                map_flag,
-                api_data,
-                map_info.name,
-                drawables,
+        # Load the map - so we can access it with get_map_v1
+        create_map = False
+        if storage_map is None:
+            # Only get the map data on startup if a) we haven't added the entity before
+            # b) The entity does not have the needed restore data.
+            if map_flag != cur_map:
+                # Only change the map and sleep if we have multiple maps.
+                await coord.api.send_command(RoborockCommand.LOAD_MULTI_MAP, [map_flag])
+                coord.current_map = map_flag
+                # We cannot get the map until the roborock servers fully process the
+                # map change.
+                await asyncio.sleep(MAP_SLEEP)
+            # Get the map data
+            map_update = await asyncio.gather(
+                *[coord.cloud_api.get_map_v1(), coord.get_rooms()],
+                return_exceptions=True,
             )
+            # If we fail to get the map, we should set it to empty byte,
+            # still create it, and set it as unavailable.
+            api_data = map_update[0] if isinstance(map_update[0], bytes) else b""
+            create_map = True
+        else:
+            api_data = storage_map
+        roborock_map = RoborockMap(
+            unique_id,
+            coord,
+            map_flag,
+            api_data,
+            map_info.name,
+            roborock_storage,
+            create_map,
+            drawables,
         )
+        entities.append(roborock_map)
+        if create_map and roborock_map.cached_map != b"":
+            storage_updates.append((map_info.name, roborock_map.cached_map))
+    hass.async_create_background_task(
+        roborock_storage.async_save_maps(storage_updates),
```

Comment:
Is the "bulk save" API worth it? The book keeping might be a little simpler if each map is saved separately.

Reply:
I am not sure. @balloob thoughts?

Reply:
OK, don't want to ping pong you back and forth here.